File: Backend/src/models/announcement.model.ts
import mongoose, { Document, Schema } from 'mongoose';
import { AppRole } from '../constants/roles';
import {
  AudienceScope,
  AnnouncementStatus,
  NotificationCategory,
  NotificationPriority,
} from '../constants/notification.constants';

export interface IAnnouncement extends Document {
  collegeId: mongoose.Types.ObjectId;
  departmentId?: mongoose.Types.ObjectId | null;
  title: string;
  body: string;
  category: NotificationCategory;
  audienceScope: AudienceScope;
  targetRole?: AppRole | null;
  targetCourseId?: mongoose.Types.ObjectId | null;
  targetSemesterId?: mongoose.Types.ObjectId | null;
  targetSectionId?: mongoose.Types.ObjectId | null;
  targetUserId?: mongoose.Types.ObjectId | null;
  publishAt: Date;
  expiresAt?: Date | null;
  priority: NotificationPriority;
  isPinned: boolean;
  status: AnnouncementStatus;
  createdBy: mongoose.Types.ObjectId;
  publishedAt?: Date | null;
  recipientCount: number;
  createdAt: Date;
  updatedAt: Date;
}

const AnnouncementSchema = new Schema<IAnnouncement>(
  {
    collegeId: { type: Schema.Types.ObjectId, ref: 'College', required: true, index: true },
    departmentId: { type: Schema.Types.ObjectId, ref: 'Department', default: null, index: true },
    title: { type: String, required: true, trim: true },
    body: { type: String, required: true, trim: true },
    category: {
      type: String,
      enum: Object.values(NotificationCategory),
      default: NotificationCategory.ANNOUNCEMENT,
      index: true,
    },
    audienceScope: {
      type: String,
      enum: Object.values(AudienceScope),
      required: true,
      index: true,
    },
    targetRole: {
      type: String,
      enum: Object.values(AppRole),
      default: null,
    },
    targetCourseId: { type: Schema.Types.ObjectId, ref: 'Course', default: null },
    targetSemesterId: { type: Schema.Types.ObjectId, ref: 'Semester', default: null },
    targetSectionId: { type: Schema.Types.ObjectId, ref: 'Section', default: null },
    targetUserId: { type: Schema.Types.ObjectId, ref: 'User', default: null },
    publishAt: { type: Date, default: Date.now },
    expiresAt: { type: Date, default: null },
    priority: {
      type: String,
      enum: Object.values(NotificationPriority),
      default: NotificationPriority.NORMAL,
    },
    isPinned: { type: Boolean, default: false },
    status: {
      type: String,
      enum: Object.values(AnnouncementStatus),
      default: AnnouncementStatus.DRAFT,
      index: true,
    },
    createdBy: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    publishedAt: { type: Date, default: null },
    recipientCount: { type: Number, default: 0 },
  },
  {
    timestamps: true,
    toJSON: {
      virtuals: true,
      transform: (_: unknown, ret: Record<string, any>) => {
        ret.id = ret._id?.toString();
        if (ret.collegeId) ret.collegeId = ret.collegeId.toString();
        if (ret.departmentId) ret.departmentId = ret.departmentId.toString();
        if (ret.targetCourseId) ret.targetCourseId = ret.targetCourseId.toString();
        if (ret.targetSemesterId) ret.targetSemesterId = ret.targetSemesterId.toString();
        if (ret.targetSectionId) ret.targetSectionId = ret.targetSectionId.toString();
        if (ret.targetUserId) ret.targetUserId = ret.targetUserId.toString();
        if (ret.createdBy) ret.createdBy = ret.createdBy.toString();
        delete ret.__v;
        return ret;
      },
    },
  }
);

// Compound indexes for fast filtered feed retrieval and tenant security
AnnouncementSchema.index({ collegeId: 1, status: 1, publishAt: -1 });
AnnouncementSchema.index({ collegeId: 1, audienceScope: 1, status: 1 });
AnnouncementSchema.index({ departmentId: 1, status: 1, publishAt: -1 });

export const Announcement = mongoose.model<IAnnouncement>('Announcement', AnnouncementSchema);

File: Frontend/lib/features/notifications/domain/models/announcement_model.dart
import '../../../auth/domain/models/role_enum.dart';
import 'notification_models.dart';

enum AnnouncementAudienceScope {
  college,
  department,
  course,
  semester,
  section,
  role,
  individual;

  String get label {
    switch (this) {
      case AnnouncementAudienceScope.college:
        return 'College-wide';
      case AnnouncementAudienceScope.department:
        return 'Department';
      case AnnouncementAudienceScope.course:
        return 'Course';
      case AnnouncementAudienceScope.semester:
        return 'Semester';
      case AnnouncementAudienceScope.section:
        return 'Section';
      case AnnouncementAudienceScope.role:
        return 'Role-based';
      case AnnouncementAudienceScope.individual:
        return 'Individual';
    }
  }

  String get apiValue => name.toUpperCase();
  String get displayName => label;

  static AnnouncementAudienceScope fromString(String? val) {
    if (val == null) return AnnouncementAudienceScope.college;
    final lower = val.toLowerCase();
    return AnnouncementAudienceScope.values.firstWhere(
      (e) => e.name.toLowerCase() == lower,
      orElse: () => AnnouncementAudienceScope.college,
    );
  }
}

enum AnnouncementStatus {
  draft,
  published,
  archived;

  String get label {
    switch (this) {
      case AnnouncementStatus.draft:
        return 'Draft';
      case AnnouncementStatus.published:
        return 'Published';
      case AnnouncementStatus.archived:
        return 'Archived';
    }
  }

  String get apiValue => name.toUpperCase();
  String get displayName => label;

  static AnnouncementStatus fromString(String? val) {
    if (val == null) return AnnouncementStatus.draft;
    final lower = val.toLowerCase();
    return AnnouncementStatus.values.firstWhere(
      (e) => e.name.toLowerCase() == lower,
      orElse: () => AnnouncementStatus.draft,
    );
  }
}

class AnnouncementModel {
  final String id;
  final String collegeId;
  final String? departmentId;
  final String title;
  final String body;
  final String category;
  final AnnouncementAudienceScope audienceScope;
  final AppRole? targetRole;
  final String? targetCourseId;
  final String? targetSemesterId;
  final String? targetSectionId;
  final String? targetUserId;
  final DateTime publishAt;
  final DateTime? expiresAt;
  final NotificationPriority priority;
  final bool isPinned;
  final AnnouncementStatus status;
  final String createdBy;
  final DateTime? publishedAt;
  final int recipientCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  const AnnouncementModel({
    required this.id,
    required this.collegeId,
    this.departmentId,
    required this.title,
    required this.body,
    this.category = 'announcement',
    required this.audienceScope,
    this.targetRole,
    this.targetCourseId,
    this.targetSemesterId,
    this.targetSectionId,
    this.targetUserId,
    required this.publishAt,
    this.expiresAt,
    this.priority = NotificationPriority.normal,
    this.isPinned = false,
    this.status = AnnouncementStatus.draft,
    required this.createdBy,
    this.publishedAt,
    this.recipientCount = 0,
    required this.createdAt,
    required this.updatedAt,
  });

  bool get isPublished => status == AnnouncementStatus.published;
  bool get isDraft => status == AnnouncementStatus.draft;
  bool get isArchived => status == AnnouncementStatus.archived;
  bool get isExpired => expiresAt != null && DateTime.now().isAfter(expiresAt!);

  factory AnnouncementModel.fromJson(Map<String, dynamic> json) {
    final rawId = json['id'] ?? json['_id'] ?? '';
    final rawCollegeId = json['collegeId'] ?? '';
    final rawCreatedBy = json['createdBy'] ?? '';

    return AnnouncementModel(
      id: rawId.toString(),
      collegeId: rawCollegeId.toString(),
      departmentId: json['departmentId']?.toString(),
      title: json['title']?.toString() ?? '',
      body: json['body']?.toString() ?? '',
      category: json['category']?.toString() ?? 'announcement',
      audienceScope: AnnouncementAudienceScope.fromString(json['audienceScope']?.toString()),
      targetRole: json['targetRole'] != null
          ? AppRole.values.firstWhere(
              (e) => e.value.toLowerCase() == json['targetRole'].toString().toLowerCase(),
              orElse: () => AppRole.student,
            )
          : null,
      targetCourseId: json['targetCourseId']?.toString(),
      targetSemesterId: json['targetSemesterId']?.toString(),
      targetSectionId: json['targetSectionId']?.toString(),
      targetUserId: json['targetUserId']?.toString(),
      publishAt: json['publishAt'] != null
          ? DateTime.tryParse(json['publishAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
      expiresAt: json['expiresAt'] != null ? DateTime.tryParse(json['expiresAt'].toString()) : null,
      priority: NotificationPriority.values.firstWhere(
        (e) => e.name.toLowerCase() == (json['priority']?.toString().toLowerCase() ?? ''),
        orElse: () => NotificationPriority.normal,
      ),
      isPinned: json['isPinned'] == true,
      status: AnnouncementStatus.fromString(json['status']?.toString()),
      createdBy: rawCreatedBy.toString(),
      publishedAt: json['publishedAt'] != null ? DateTime.tryParse(json['publishedAt'].toString()) : null,
      recipientCount: (json['recipientCount'] as num?)?.toInt() ?? 0,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.tryParse(json['updatedAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'collegeId': collegeId,
      'departmentId': departmentId,
      'title': title,
      'body': body,
      'category': category,
      'audienceScope': audienceScope.apiValue,
      'targetRole': targetRole?.value,
      'targetCourseId': targetCourseId,
      'targetSemesterId': targetSemesterId,
      'targetSectionId': targetSectionId,
      'targetUserId': targetUserId,
      'publishAt': publishAt.toIso8601String(),
      'expiresAt': expiresAt?.toIso8601String(),
      'priority': priority.name,
      'isPinned': isPinned,
      'status': status.apiValue,
      'createdBy': createdBy,
      'publishedAt': publishedAt?.toIso8601String(),
      'recipientCount': recipientCount,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  AnnouncementModel copyWith({
    String? id,
    String? collegeId,
    String? departmentId,
    String? title,
    String? body,
    String? category,
    AnnouncementAudienceScope? audienceScope,
    AppRole? targetRole,
    String? targetCourseId,
    String? targetSemesterId,
    String? targetSectionId,
    String? targetUserId,
    DateTime? publishAt,
    DateTime? expiresAt,
    NotificationPriority? priority,
    bool? isPinned,
    AnnouncementStatus? status,
    String? createdBy,
    DateTime? publishedAt,
    int? recipientCount,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return AnnouncementModel(
      id: id ?? this.id,
      collegeId: collegeId ?? this.collegeId,
      departmentId: departmentId ?? this.departmentId,
      title: title ?? this.title,
      body: body ?? this.body,
      category: category ?? this.category,
      audienceScope: audienceScope ?? this.audienceScope,
      targetRole: targetRole ?? this.targetRole,
      targetCourseId: targetCourseId ?? this.targetCourseId,
      targetSemesterId: targetSemesterId ?? this.targetSemesterId,
      targetSectionId: targetSectionId ?? this.targetSectionId,
      targetUserId: targetUserId ?? this.targetUserId,
      publishAt: publishAt ?? this.publishAt,
      expiresAt: expiresAt ?? this.expiresAt,
      priority: priority ?? this.priority,
      isPinned: isPinned ?? this.isPinned,
      status: status ?? this.status,
      createdBy: createdBy ?? this.createdBy,
      publishedAt: publishedAt ?? this.publishedAt,
      recipientCount: recipientCount ?? this.recipientCount,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
